### ISSUE TITLE
[Module Proposal]: `avm-res-network-privatednszone` (tf)

### ISSUE THREAD

**Issue by chianw** [Azure/Azure-Verified-Modules#306]
### Check for previous/existing GitHub issues/module proposals

- [X] I have checked for previous/existing GitHub issues/module proposals.

### Check this module doesn't already exist in the module indexes

- [X] I have checked for that this module doesn't already exist in the module indexes; or I'm proposing the module to be migrated from CARML/TFVM.

### Bicep or Terraform?

Terraform

### Module Classification?

Resource Module

### Module Name

avm-res-network-privatednszone

### Module Details

Resource module for private DNS zone which should support A, AAAA, CNAME, MX, PTR, SRV, TXT records and virtual network link.

### Do you want to be the owner of this module?

Yes

### Module Owner's GitHub Username (handle)

chianw

**Comment by mbilalamjad** [comment]
Hi @chianw,

Thanks for requesting/proposing to be an AVM module owner!

We just want to confirm **you agree to the below pages** that define what module ownership means:

- [Team Definitions & RACI](https://azure.github.io/Azure-Verified-Modules/specs/shared/team-definitions)
- [Shared Specification (Bicep & Terraform)](https://azure.github.io/Azure-Verified-Modules/specs/shared)
- [Module Support](https://azure.github.io/Azure-Verified-Modules/help-support/module-support)

Any questions or clarifications needed, let us know!

If you agree, please just **reply to this issue with the exact sentence below** (as this helps with our automation 👍):

"I CONFIRM I WISH TO OWN THIS AVM MODULE AND UNDERSTAND THE REQUIREMENTS AND DEFINITION OF A MODULE OWNER"

Thanks,

The AVM Core Team

#RR

<br>

> **NOTE**:
>
> - This Module Proposal issue **MUST remain open** until the module is fully developed, tested and published to the relevant registry. **Do NOT close** the issue before the successful publication is confirmed!
> - Once the module is fully developed, tested and published to the relevant registry, and the Module Proposal issue was closed, it **MUST remain closed**.

**Comment by chianw** [comment]
I CONFIRM I WISH TO OWN THIS AVM MODULE AND UNDERSTAND THE REQUIREMENTS AND DEFINITION OF A MODULE OWNER

**Comment by mbilalamjad** [comment]
Hi @chianw,

Thanks for confirming that you wish to own this AVM module and understand the related requirements and responsibilities!

We just want to ask you to double check a few important things and take the next steps before you start the development.

**Please use the following values explicitly as provided in the [module index](https://azure.github.io/Azure-Verified-Modules/indexes/) page**:

- For your module:
  - `ModuleName` - to name your module
  - `TelemetryIdPrefix` - to be used in your module's [telemetry](https://azure.github.io/Azure-Verified-Modules/specs/shared/#id-sfr3---category-telemetry---deploymentusage-telemetry)
- For your module's repository:
  - Repository name and folder path are defined in `RepoURL`
  - Create the GitHub teams defined in the `ModuleOwnersGHTeam` and `ModuleContributorsGHTeam` columns and grant them permissions as described [here](https://azure.github.io/Azure-Verified-Modules/specs/shared/#id-snfr20---category-contributionsupport---github-teams-only).
    - In case of a Bicep module, **please let the AVM core team know** when you created these teams, and we'll help you with assigning the necessary permissions as described [here](https://azure.github.io/Azure-Verified-Modules/specs/shared/#grant-permissions---bicep).
    - In case of a Terraform module, follow the instructions [here](https://azure.github.io/Azure-Verified-Modules/specs/shared/#grant-permissions---terraform) and assign the necessary permissions.

Grant the right level of permissions for  the AVM core team and PG teams on your GitHub repo as described [here](https://azure.github.io/Azure-Verified-Modules/specs/shared/#id-snfr9---category-contributionsupport---avm--pg-teams-github-repo-permissions).

You can now start the development of this module! ✅ Happy coding! 🎉

**Please respond to this comment and request a review from the AVM core team once your module is ready to be published!** Please include a link pointing to your PR if available. 🙏

Any further questions or clarifications needed, let us know!

Thanks,

The AVM Core Team

<br>

> **NOTE**:
>
> - This Module Proposal issue **MUST remain open** until the module is fully developed, tested and published to the relevant registry. **Do NOT close** the issue before the successful publication is confirmed!
> - Once the module is fully developed, tested and published to the relevant registry, and the Module Proposal issue was closed, it **MUST remain closed**.

**Comment by chianw** [comment]
Hi @mbilalamjad noted on this, I've created the necessary GH owner and contributor teams as well as the repo in Azure organization. Will start to work on it.

**Comment by chianw** [comment]
Hi @mbilalamjad I have created the module and did testing successfully, however it is still a private repository at https://github.com/Azure/terraform-azurerm-avm-res-network-privatednszone . The team managing this is not allowing the repository to become a public because the support statement from https://github.com/Azure/terraform-azurerm-avm-template/blob/main/SUPPORT.md is a generic one and this is the template from which the repo is based on. 

The request from the team managing this is that the AVM core team/PM to agree on the standard support.md for all AVM modules so that I can get on to the next stage which is legal review.

![image](https://github.com/Azure/Azure-Verified-Modules/assets/91816369/f4de5ada-25ef-47e0-b689-66a9f7b01cb9)

**Comment by mbilalamjad** [comment]
Thanks @chianw for raising awareness about the above, I've asked @PmeshramPM and @matt-FFFFFF to advise on this as I'm out of office till 2nd Jan

**Comment by mbilalamjad** [comment]
In the meantime, could you add the avm-core-team, avm-res-network-privatednszone-module-contributor-tf & avm-res-network-privatednszone-module-owner-tf like below for us to review the module.

![image](https://github.com/Azure/Azure-Verified-Modules/assets/79409563/37cb7363-26b5-4e0c-9a79-3639091cc9a0)

**Comment by chianw** [comment]
Hi @mbilalamjad apologies for the delay as I was out of office. I've added avm-core-team as admin role to the repository. The code is in a branch called **pre-release1**

**Comment by mbilalamjad** [comment]
Hey @chianw, hope you're doing great and happy new year, I've resumed work and will do a review of the module next week.

**Comment by chianw** [comment]
Hi @mbilalamjad hope you are well. Have you had a chance to review the pre-release1 branch of the repository? I've tried to update the README.md file in the pre-release1 branch and did a PR to merge into main but it's failing with some errors below. Let me know how I can fix the errors below, we can connect for a quick session if that helps

![image](https://github.com/Azure/Azure-Verified-Modules/assets/91816369/e37486e3-40d2-472b-8d1e-428c02338316)

**Comment by lonegunmanb** [comment]
Hi @chianw you can check linting error by clicking `Details`, many AVM modules do not meet the AVM's spec now and we've added a new TFLint plugin to enforce these rules. If you don't know how to fix s specific tflint issue, you can share it here and ping me, I'm glad to help you.

**Comment by chianw** [comment]
Hi @lonegunmanb thanks for reaching out. I looked at the linting error details below, but I don't understand what it is looking for. Able to point in the right direction?

![image](https://github.com/Azure/Azure-Verified-Modules/assets/91816369/46b40044-4de4-4d8d-8d44-3eb9ee056926)

**Comment by chianw** [comment]
Hi @mbilalamjad hope you are well, are you able to point me in the right direction about the linting errors?

**Comment by mbilalamjad** [comment]
@chianw I hope @segraef was able to help on the internal chat, let us know if there's further support needed.

**Comment by chianw** [comment]
@mbilalamjad I'm trying to arrange a session with @segraef to see if he can guide me on fixing the linting errors

**Comment by segraef** [comment]
@chianw I thin this is solved right?

**Comment by mbilalamjad** [comment]
Thank you so much @chianw for owning and developing this module.

Closing this proposal as module has been published to terraform registry below.
https://registry.terraform.io/modules/Azure/avm-res-network-privatednszone/azurerm/latest

Further discussion on this module including issues & PRs to take please on its repo below.
https://github.com/Azure/terraform-azurerm-avm-res-network-privatednszone

**Comment by jeanchg** [comment]
this module ownership is being transitioned to @MinHeinA
